viadot/sources/sap_rfc.py

```diff
@@ -463,37 +468,86 @@ def to_df(self):
         else:
             SEPARATORS = [sep]
 
-        records = None
         for sep in SEPARATORS:
             logger.info(f"Checking if separator '{sep}' works.")
-            df = pd.DataFrame()
+            df = pd.DataFrame(columns=columns)
             self._query["DELIMITER"] = sep
             chunk = 1
+            row_index = 0
             for fields in fields_lists:
                 logger.info(f"Downloading {chunk} data chunk...")
+                self._query["FIELDS"] = fields
                 try:
-                    self._query["FIELDS"] = fields
-                    try:
-                        response = self.call(func, **params)
-                    except ABAPApplicationError as e:
-                        if e.key == "DATA_BUFFER_EXCEEDED":
-                            raise DataBufferExceeded(
-                                "Character limit per row exceeded. Please select fewer columns."
-                            )
-                        else:
-                            raise e
-                    record_key = "WA"
-                    data_raw = response["DATA"]
-                    records = [row[record_key].split(sep) for row in data_raw]
-                    df[fields] = records
-                    chunk += 1
-                except ValueError:
-                    df = pd.DataFrame()
-                    continue
-        if not records:
-            logger.warning("Empty output was generated.")
-            columns = []
-        df.columns = columns
+                    response = self.call(func, **params)
+                except ABAPApplicationError as e:
+                    if e.key == "DATA_BUFFER_EXCEEDED":
+                        raise DataBufferExceeded(
+                            "Character limit per row exceeded. Please select fewer columns."
+                        )
+                    else:
+                        raise e
+                record_key = "WA"
+                data_raw = np.array(response["DATA"])
+
+                if row_index == 0:
+                    row_index = data_raw.shape[0]
+                    if row_index == 0:
+                        logger.warning(
+                            f"Empty output was generated for chunk {chunk} in columns {fields}."
+                        )
+                        chunk += 1
+                        continue
+                elif data_raw.shape[0] != row_index:
+                    data_raw = data_raw[:row_index]
+                    logger.warning(
+                        f"New rows were generated during the execution of the script. The table is truncated to the number of rows for the first chunk"
+                    )
+
+                # first we identify where the data has an extra sep due to text
+                sep_counts = np.array([], dtype=int)
```

Comment:
I don't think `numpy` is needed here or below either, seems like you can do that with regular Python lists.

But anyway take a step back and think if this problem can be solved using escape characters since you're adding a lot of custom logic here just to basically do that. 

Also, please split the logic into functions so that it's easier to follow (and also add docstrings to those functions explaining what each of them is doing + typing). Looking at the code below it should probably be ~5 functions. The for loop as it is now has ~80 lines of code; it shouldn't be more than 10-20.

Once you rewrite (or use escape character approach) I'll be able to provide some feedback on the logic.

Reply:
done, have a look at it.